File: map-reduce/src/main/scala/by/bsuir/labs/akka/scala/actors/ReduceActor.scala
package by.bsuir.labs.akka.scala.actors

import java.util.ArrayList
import java.util.HashMap

import scala.collection.JavaConversions.asScalaBuffer

import akka.actor.actorRef2Scala
import akka.actor.Actor
import akka.actor.ActorRef
import by.bsuir.labs.akka.scala.{ReduceData, MapData, Word}

class ReduceActor(aggregateActor: ActorRef) extends Actor {

  def receive: Receive = {
    case message: MapData =>
      aggregateActor ! reduce(message.dataList)
    case _ =>

  }

  def reduce(dataList: ArrayList[Word]): ReduceData = {
    val reducedMap = new HashMap[String, java.lang.Integer]
    for (wc: Word <- dataList) {
      val word: String = wc.word
      if (reducedMap.containsKey(word)) {
        reducedMap.put(word, reducedMap.get(word) + 1)
      } else {
        reducedMap.put(word, 1)
      }
    }
    new ReduceData(reducedMap)
  }
}

File: map-reduce/src/main/java/by/bsuir/labs/akka/actor/ReduceActor.java
package by.bsuir.labs.akka.actor;

import java.util.HashMap;
import java.util.List;

import akka.actor.ActorRef;
import akka.actor.UntypedActor;
import by.bsuir.labs.akka.message.MapData;
import by.bsuir.labs.akka.message.ReduceData;
import by.bsuir.labs.akka.message.WordCount;

public class ReduceActor extends UntypedActor {

	private ActorRef aggregateActor = null;

	public ReduceActor(ActorRef aggregateActorRef) {
		aggregateActor = aggregateActorRef;
	}

	@Override
	public void onReceive(Object message) throws Exception {
		if (message instanceof MapData) {
			MapData mapData = (MapData) message;
			ReduceData reduceData = reduce(mapData.getDataList());
			aggregateActor.tell(reduceData);
		} else
			unhandled(message);
	}

	private ReduceData reduce(List<WordCount> dataList) {
		HashMap<String, Integer> reducedMap = new HashMap<String, Integer>();
		for (WordCount wordCount : dataList) {
			if (reducedMap.containsKey(wordCount.getWord())) {
				Integer value = reducedMap.get(wordCount.getWord());
				value++;
				reducedMap.put(wordCount.getWord(), value);
			} else {
				reducedMap.put(wordCount.getWord(), 1);
			}
		}
		return new ReduceData(reducedMap);
	}
}
